web/stream/pool.go

```diff
@@ -0,0 +1,86 @@
+package stream
+
+import (
+	"errors"
+	"sync"
+)
+
+// Pool is a pool of channels.
+// The channels are used to send data to connected clients.
+type Pool[T any] struct {
+	mu sync.Mutex
+	// Map of channels that are used to send data to the client.
+	// The key is the user ID.
+	channels map[uint64]chan *T
+}
+
+// NewPool creates a new pool.
+func NewPool[T any]() *Pool[T] {
+	return &Pool[T]{
+		channels: make(map[uint64]chan *T),
+	}
+}
+
+// Add adds a new channel to the pool.
+func (p *Pool[T]) Add(id uint64) chan *T {
+	// Delete the channel if it already exists.
+	p.Remove(id)
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	// Create a new channel.
+	p.channels[id] = make(chan *T)
+	return p.channels[id]
+}
+
+// Remove removes a channel from the pool.
+func (p *Pool[T]) Remove(id uint64) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if ch, ok := p.channels[id]; ok {
```

Comment:
Since this is called from `Add`, before taking the lock in `Add`, it will take the lock here in `Remove` and release it again. Then in `Add` it will take the lock again. Not sure this is what you want, leaving to another goroutine to grab the lock in-between. Also, it adds pressure on the lock.

Suggested alternative; implement unexported `remove` without locking. Then require that callers of `remove` must already hold the lock before calling.

```go
func (p *Pool[T]) remove(id uint64) {
	if ch, ok := p.channels[id]; ok {
		close(ch)
 	}
 	delete(p.channels, id)
```

Reply:
That is, `Remove` (exported) and `Add` must hold the lock before calling `remove` (unexported).

Reply:
If we don't want to export `Remove`, we should call it `remove`, and the other one should be called `internalRemove` to signal that we should be used with care (and documentation should explain that it requires holding the lock.)

Another possible name is `noLockRemove`.